---
title: "breakclone: Estimating clonal relatedness probability between tumour pairs"
author: "Maria Roman Escorza (maria.roman-escorza@kcl.ac.uk)"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    number_sections: true
    self_contained: yes
    css: corp-styles.css
    highlight: pygments
vignette: >
  %\VignetteIndexEntry{breakclone: Estimating clonal relatedness probability between tumour pairs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction 

When a cancer patient develops a new tumor, it's crucial to determine whether the new tumor is a recurrence (metastasis) of the original cancer or a completely new instance of the disease. Since each tumor carries a unique profile of somatic genomic alterations, it is theoretically possible to provide a more definitive diagnosis in these difficult cases by comparing the genomic patterns of the two tumors.

`breakclone` explores the genomic similarity and clonal relationship between pairs of tumor samples by comparing their genomic profiles, using either copy number or mutation data. Unlike other tools (ref), breakclone incorporates both population-level allele frequencies and sample-specific allele frequencies when analyzing mutation data to assess clonal relatedness. For copy number data, it focuses on individual copy number aberration breakpoints, rather than chromosome-arm-level events, to establish clonal connections. This approach is corrected for the frequency of such events within the cohort. Although this method can be more challenging to apply across cohorts analyzed using different techniques, it provides stronger evidence of clonal relationships when shared breakpoints are found between lesions. A reference distribution of concordance scores is generated by randomly permuting pairs of tumors from different patients. The number of permutations is determined empirically to ensure convergence of the distribution and is used to calculate the p-values for the concordance score of each tumor pair.


## Citation

If you find this tool useful, please cite:

------------------------------------------------------------------------

***Lips EH, Kumar T, Megalios A, et al. 2022. Genomic analysis defines clonal relationships of ductal carcinoma in situ and recurrent invasive breast cancer. [Nature Genetics](https://doi.org/10.1038/s41588-022-01082-3)***

------------------------------------------------------------------------

# Installation

```{r, eval=FALSE}
if (!require("devtools")) {
  install.packages("devtools")
}

devtools::install_github("Sawyer-s-Group/breakclone")
```


# Overview of the package

`breakclone` functions can be categorized into mainly Reading, Visualization and Analysis modules. breakclone can be run in three modes depending on your data: `alleleSpecific` for the allele-specific copy number data, `VCF` for the non-allele specific copy number data and **mutation** data. Each of these functions and a short description is summarized as shown below. 

![](overview.png)

We load the package and the datasets that will be used in this vignette:

```{r, warning=FALSE}
library(breakclone)
data(segmentTable_AS)
data(segmentTable)
data(mutationTable)
```

# Preparing input data

`breakclone` takes advantage of fast `data.table`, an extension of `data.frame` which allows fast aggregation of large data. If you prefer to use alternative methods to read your data than the proposed by the package, please make sure you transform it to the appropiate formate using `data.table::as.data.table()`.

Minimal inputs for running brekclone are:

- Aberrations `data.table`  (called `segmentTable` in copy number functions and `mutationTable`).
- paired IDs `data.table` (`pairs`). 

## `segmentaTable` input for non-allele-specific data 

`readVCFCn` automatically reads copy number data from standard VCF files. 

```{r, eval=FALSE}
segmentTable <- readVCFCn("/path/to/data")
```

```{r}
class(segmentTable)
head(segmentTable)
```

## `segmentTable` input for allele-specific data

`readAlleleSpecific` automatically reads text files with columns for sample IDs, chromosome, start, end, number of probes, number of copies in the major and minor allele.

```{r, eval=FALSE}
segmentTable_AS <- readAlleleSpecific("/path/to/data")
```

```{r}
class(segmentTable_AS)
head(segmentTable_AS)
```

## `mutationTable` input

`readVCFMutations` automatically reads mutations from standard VCF files. It is recommended to include annotation for visualization purposes.

```{r, eval=FALSE}
mutationTable <- readVCFMutations("/path/to/data")
```

```{r}
class(mutationTable)
head(mutationTable)
```

## `pairs` input

`pairs` is a table of paired samples from the dataset. All tumours present in this table will be paired with all tumours from other patients. IIf sample IDs comprise a patient ID and a tumour type identifier, `inferPairs` can attempt to infer the tumour pairs from the cohort. 

```{r}
pairs_cn <- inferPairs(segmentTable)
pairs_cnAS <- inferPairs(segmentTable_AS)
pairs_muts <- inferPairs(mutationTable)
head(pairs_muts)
```

# Clonality assessment using copy number  

Firstly, `breakclone` generates the reference distribution of concordance scores from unpaired tumours for a given dataset. As distribution of concordance scores is calculated by randomly permuting all possible pairs from different patients, we set a seed to be able to reproduce the results of this vignette.

```{r}
set.seed(111)
reference <- makeReferenceCN(
  segmentTable = segmentTable,
  pairs = pairs_cn,
  cnType = "VCF"
)
```

We are ready to run the clonality assessment using `calculateRelatednessCn` which calculates the relatedness scores and p-values for paired tumours from copy number data. Clonality results of copy number data can be summarized using `summarizeClonalityCN`. By default, threshold for ambiguous clonality is 0.05 and for related is 0.01. 

```{r}
results <- calculateRelatednessCn(
  segmentTable = segmentTable,
  pairs = pairs_cn,
  reference = reference,
  cnType = "VCF"
)

summary <- summarizeClonalityCN(
  clonalityResults = results,
  segmentTable,
  thres_ambiguous = 0.05,
  thres_related = 0.01,
  cnType = "VCF"
)

head(summary)
```

The algorithm uses the position of the individual copy number aberration breakpoints. Concordant breakpoints can be exported using `getSharedBreaks`. Output is a list of tables including he shared breakpoints per sample.

```{r}
breaks <- getSharedBreaks(
  segmentTable = segmentTable,
  pairs = pairs_cn,
  cnType = "VCF",
  save = FALSE
)
breaks[[1]]
```

Paired copy number profiles with shared CNAs and breakpoints can be visualized using `plotCNpairVCF`. Input is binned copy number data where  `cnTable` represents the copy number values and `binnedTable` represents the segment mean of the copy number values, which can be extracted easily from a segmented `QDNAseq` object using the code below.

```{r, eval=FALSE}
binnedTable <- log2(Biobase::assayDataElement(QDNAseq_object, "segmented"))
cnTable <- log2(Biobase::assayDataElement(QDNAseq_object, "copynumber"))
```


```{r, message=FALSE, warning=FALSE}
data(binnedTable)
data(cnTable)

plotCNpairVCF(binnedTable,
  cnTable,
  pair = c("Patient53_Primary", "Patient53_Recurrence"),
  segmentTable = segmentTable,
  breaks = breaks,
  build = "hg38"
)
```

# Clonality assessment using allele-specific copy number  

```{r, warning=FALSE, message=FALSE}
reference <- makeReferenceCN(
  segmentTable = segmentTable_AS,
  pairs = pairs_cnAS
)
```

```{r, message=FALSE, warning=FALSE}
results <- calculateRelatednessCn(
  segmentTable = segmentTable_AS,
  pairs = pairs_cnAS,
  reference = reference
)

summary <- summarizeClonalityCN(
  clonalityResults = results,
  segmentTable = segmentTable_AS
)
head(summary)
```


```{r, message=FALSE, warning=FALSE}
breaks <- getSharedBreaks(
  segmentTable = segmentTable_AS,
  pairs = pairs_cnAS,
  save = FALSE
)
```

# Clonality assessment using mutation

For the clonality assessment using mutation data, the allele frequency is weighted according to the population frequency. A higher allele frequency is interpreted as a stronger indicator of clonal relatedness, while a higher population frequency is interpreted as diminishing the predictive value of the variant. To calculate the reference distribution of concordance scores, the TCGA Pan-Cancer Atlas breast cancer mutation calls were used available in the package as `brca`. 

```{r}
data(brca)
head(brca)
```

```{r, warning=FALSE}
reference <- makeReferenceMutations(
  mutationTable = mutationTable,
  pairs = pairs_muts,
  additionalMutations = brca,
  nAdditionalSamples = 346
)
```

We are ready to run the clonality assessment using `calculateRelatednessMutations` which calculates the relatedness scores and p-values for paired tumours from mutation data. Results can be summarized using `summarizeClonalityMuts`. 

```{r, warning=FALSE}
results <- calculateRelatednessMutations(
  mutationTable = mutationTable,
  pairs = pairs_muts,
  reference = reference,
  additionalMutations = brca,
  nAdditionalSamples = 346
)

summary <- summarizeClonalityMuts(
  clonalityResults = results,
  mutationTable = mutationTable
)
head(summary)
```

Shared mutations can be exported using `getSharedMuts` and plotted with `plotScatterVAF`. Mutations can be labelled by extra column in `mutationTable` if `annotGenes=TRUE`.

```{r}
shared_muts <- getSharedMuts(
  mutationTable = mutationTable,
  pairs = pairs_muts,
  save = FALSE
)
shared_muts[[4]]

plotScatterVAF(
  mutationTable = mutationTable,
  pair = c("Patient46_Primary", "Patient46_Recurrence"),
  annotGenes = TRUE
)
```


# Assessement of the clonality score

`plotScoresDensity` plots the reference and pair distributions and `plotScoresHistogram` plots the reference and pair histograms.

```{r}
plotScoresDensity(reference, results)
```

```{r, warning=FALSE}
plotScoresHistogram(reference, results)
```

# Plot clonality results

You can plot clonality results using `plotSummary`. An extra pheno bar can be added using `extraAnno`.

```{r, message=FALSE}
plotSummary(summary,
  extraAnno = rep("Invasive", nrow(summary)),
  colorsExtraAnno = c(
    "DCIS" = "#e6cde3",
    "Invasive" = "#af67a7"
  )
)
```

# Appendix: Download mutations from TCGA 

This section shows how to create `additionalMutations`, a table of mutations to be taken into account when calculating population frequencies. At a minimum, a table of the mutations in the population being studied. This is more informative when tumour type-specific mutations are included from external sources (e.g. TCGA).

```{r}
library(TCGAretriever)

brca <- get_mutation_data(
  case_list_id = "brca_tcga_pub_complete",
  gprofile_id = "brca_tcga_pub_mutations",
  glist = unique(mutationTable$annotation)
)
brca$chr <- sub("23", "X", brca$chr) # 23 to X
brca$chr <- paste0("chr", brca$chr)
brca <- brca[brca$startPosition == brca$endPosition, ] # take only SNPs mutations
brca <- brca[brca$mutationType %in% c("Missense_Mutation", "Nonsense_Mutation", "Nonstop_Mutation", "Splice_Region", "Splice_Site", "Translation_Start_Site"), ] # filter specific mutations
```

We downloaded BRCA TCGA mutations from `r length(unique(brca$sampleId))` samples using `TCGAretriever` package. `additionalMutations` needs to be transform to a `GRanges` object.

```{r, eval=FALSE}
brca <- makeGRangesFromDataFrame(brca,
  start.field = "startPosition",
  end.field = "endPosition",
  seqnames.field = "chr"
)
```

As the dataset used in this vignette was aligned to hg38 and TCGA to hg19, I liftover the mutations:

```{r, eval=FALSE}
library(liftOver)

download.file(
  "https://hgdownload.soe.ucsc.edu/gbdb/hg19/liftOver/hg19ToHg38.over.chain.gz",
  "hg19ToHg38.over.chain.gz"
)
R.utils::gunzip("hg19ToHg38.over.chain.gz", overwrite = TRUE)

ch <- import.chain("hg19ToHg38.over.chain")
brca <- unlist(liftOver(brca, ch))
```


# Session information

```{r}
sessionInfo()
```

# References

Remember to include any relevant references to methods.
